locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/AbstractComputeServiceRegistry.java

```diff
@@ -0,0 +1,283 @@
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
+import com.google.inject.Module;
+
+public abstract class AbstractComputeServiceRegistry implements ComputeServiceRegistry, JcloudsLocationConfig {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractComputeServiceRegistry.class);
+
+    private final Map<Map<?, ?>, ComputeService> cachedComputeServices = new ConcurrentHashMap<>();
+
+    @Override
+    public ComputeService findComputeService(ConfigBag conf, boolean allowReuse) {
+        PropertiesBuilder propertiesBuilder = new PropertiesBuilder(conf)
+                .setCommonProperties();
+
+        Iterable<Module> modules = getCommonModules();
+
+        // Enable aws-ec2 lazy image fetching, if given a specific imageId; otherwise customize for specific owners; or all as a last resort
+        // See https://issues.apache.org/jira/browse/WHIRR-416
+        String provider = getProviderFromConfig(conf);
+        if ("aws-ec2".equals(provider)) {
+            propertiesBuilder.setAWSEC2Properties();
+        } else if ("azurecompute-arm".equals(provider)) {
+            propertiesBuilder.setAzureComputeArmProperties();
+            // jclouds 2.0.0 does not include the rate limit module for Azure ARM. This quick fix enables this which will
+            // avoid provisioning to fail due to rate limit exceeded
+            // See https://issues.apache.org/jira/browse/JCLOUDS-1229
+            modules = ImmutableSet.<Module>builder()
+                    .addAll(modules)
+                    .add(new AzureComputeRateLimitModule())
+                    .build();
+        }
+
+        Properties properties = propertiesBuilder
+                .setJCloudsProperties()
+                .setEndpointProperty()
+                .build();
+
+        Supplier<ComputeService> computeServiceSupplier =  new ComputeServiceSupplier(conf, modules, properties);
+        if (allowReuse) {
+            return cachedComputeServices.computeIfAbsent(makeCacheKey(conf, properties), key -> computeServiceSupplier.get());
+        }
+        return computeServiceSupplier.get();
+    }
+
+    private Map<?, ?> makeCacheKey(ConfigBag conf, Properties properties) {
+        String provider = getProviderFromConfig(conf);
+        String identity = checkNotNull(conf.get(CloudLocationConfig.ACCESS_IDENTITY), "identity must not be null");
+        String credential = checkNotNull(conf.get(CloudLocationConfig.ACCESS_CREDENTIAL), "credential must not be null");
+        String endpoint = properties.getProperty(Constants.PROPERTY_ENDPOINT);
+        return MutableMap.builder()
+                .putAll(properties)
+                .put("provider", provider)
+                .put("identity", identity)
+                .put("credential", credential)
+                .putIfNotNull("endpoint", endpoint)
+                .build()
+                .asUnmodifiable();
+    }
+
+    public class PropertiesBuilder {
```

Comment:
Is it maybe `JcloudsPropertiesBuilder` a better name?
Can we also put it in its own class, maybe?